README
PDIC CUI search - PDICを検索とか（ほげほげとか）する簡易CUIツール
であります

(c) 2012 by naoya_t

【注意】
2012/1/22のバージョン以前に作られたTOCはPDICのindexに出てくる順番になっているため
アドレスが正しくソートされておらず、正しく出てこない単語がある可能性があります。
お手数ですが（ちょっと時間がかかりますが）.make toc で索引を作り直して下さい。

■目的

PDIC形式の辞書資産の活用（+個人的な研究）が目的です。

内部でBOCU-1符号化された上、見出し語が圧縮格納されているPDIC形式は、
データ収納効率はよく考えられていて素晴らしいと思うのですが、
そのままでは近代的で高速な全文検索手法には向いていないと思うので、
例えば英辞郎の検索だけが目当てなら、取り急ぎPDIC形式辞書ではなくTEXT形式の辞書を入手した上、
saryとかkyoto cabinetで検索したらいいんじゃないかと思います。

...というわけで、suffix arrayを用いた高速検索なんかにも対応してみました。
...GUIフロントエンドも多分そのうち作ります。

■必要なもの

- GNU C++
- iconv, gtest, re2
- PDIC辞書がないと意味ない

■簡単な用法

.pdicrc.sample を ~/.pdicrc としてコピーして、編集して使ってください。

起動するとプロンプトが出ます。辞書が選択されていれば、入力した文字列を検索します。正規表現でいうと
  (1) word → /^word$/
  (2) word* → /^word/
  (3) *word → /word/
  (4) /word/ → /word/ (※そのまま正規表現として)
  (5) +word → /word/
のような検索になります。

(1)(2)ではPDICが持つインデックスを利用した前方一致検索を行います。

(3) の*word ではsuffix arrayを用いて高速に見出し語全検索を行います。（要インデックス作成）
※今はインデックスを小さめにしていて、単語の先頭からしかマッチしないようになっています。英単語の途中からのマッチングをどうするか、音節単位とか、接頭辞接尾辞区切りとか検討中です。(1/12)

(4) では正規表現ライブラリre2を利用した見出し語全検索を行います。単語の一部分にマッチするような検索も可能です。
(5) は(1),(2),(3),(4)の和集合を得るのですが、大抵の場合(4)と同様です。

AND/OR検索は、とりあえず正規表現に置き換えるなどして何とかしてください。

readlineには（少なくとも現時点では）対応していません。（rlwrapを使ってください）

■特殊コマンド

特殊コマンドは先頭に . をつけます。

.list
    - 現在ロードされている辞書一覧。
.aliases
    - 設定されているエイリアス一覧。
.add loadpath
    - 辞書ロードパスを追加。
.group グループ名 = 名前1 名前2 ... 
    - 辞書グループを登録。
.load ファイル名
    - 辞書を
.use 辞書名
    - ターゲット辞書を変更。
.make toc
    - カレント辞書のsuffix arrayインデックスを生成します。（今のところ見出し語・訳語のみ）
.dump {header|index|words|datablock|all}
    - ダンプコマンド。〈これはいまのところ開発用〉
    - 見出し語の全リストが .dump words で取れるとか
.lookup 文字列
    - 検索実行。プロンプトに単に文字列を打ち込むのと同義。
.set 変数 = 値
    - limit = 数値
        数値: 結果表示件数の上限。デフォルトは150
    - lookup = {exact|forward|sarray|regexp|all}
        exact: デフォルトを完全一致検索に設定
        forward: 〃前方一致検索に設定
        sarray: 〃単語区切りでの一致に設定
        regexp: 〃正規表現検索に設定
        all: 〃全検索方式の和集合に設定
    - stop_on_limit = {on|off}
        on: 検索表示件数の上限に来たら検索を停止する
        off: 〃 停止しない（あとで検索結果を全部見たいときなどに用いる）
    - verbose = {on|off}
        on: メッセージ出力多め
        off: 少なめ
    - full = {on|off}
        on: 見出し語, 訳語の両方を検索対象とする（発音記号、例文にもそのうち対応）
        off: 見出し語のみ
    - direct = {on|off}
        on: 結果を（配列に格納してソートして、というステップを飛ばして）直接垂れ流す
        off: 結果を配列に格納してソートしてから出力。
             検索が終わるまで出力が始まらないので、人によってはイラッとするが
    - coloring = {on|off}
        on: 結果出力中の検索文字列を色彩＆強調表示を行う
        off: 〃行わない
    - separator = {on|off}
        on: 検索結果に辞書区切りを入れる
        off: 〃入れない
.last
    - 最後の検索結果を表示。
       last n
       last n-
       last -m
       last n-m
      のように、何番目の結果を表示するか指定することが可能。（辻褄があう範囲でなら組み合わせ可）
      この指定がある場合、limit の数値は無視される。
.quit あるいは .bye
    - 終了。インデックスファイル等をunmmapするのに数秒待つ場合があります。

■注意事項

GNU Lesser General Public License version 2 で公開します。COPYING 参照。
無保証です。
あと、辞書は付属していません。

■連絡先

@naoya_t とか


Changelog
PDIC CUI search ざっくりとした更新履歴

(c) 2012 by naoya_t

Jan 28, 2012
- 辞書データに含まれている変化形情報を索引化して活用 [英辞郎のみ]
- コードを整理
- 作業ブランチであれこれ
- 特に根拠はないけどバージョン0.7.1

Jan 23, 2012
- cpplint.py が通るように色々頑張った
- シェル部分をクラス化。Shell *g_shell 以外のグローバル変数を撲滅

Jan 22, 2012
- 検索結果をまずword_idの集合として保持するよう改造。それに伴うTOC作り直し等大工事
- コンパイル時に警告が出る箇所を手直し
- 入力が + で始まる場合にPDIC索引/suffix-array/regexp検索の和集合を得る
- 特に根拠はないけどバージョン0.7

Jan 21, 2012
- .make xml でMacOSXの辞書app向けXMLファイルが作れるようになる途中
- コンパイル時に警告が出る箇所を手直し

Jan 20, 2012
- status コマンドに各種フラグの設定値出力を追加

Jan 19, 2012
- entry_buf等のオフセットから添字を逆引きするためのmapを起動時に生成（＆終了時に破壊）していたのが割と時間がかかって嫌だったのでこれを廃止。STLのmapってどうせO(1)じゃないし。
  - で、binary searchにしても20回前後のint比較で済むのでbinary searchにした上でメモ化。検索時間はほぼそのまま、使用メモリ低減、起動＆終了にかかる時間が激減。一石二鳥。
  - 英辞郎の辞書をグループ化する際、和英辞郎を入れても余裕になった
- entry, jword, example, pron の４つ組を構造体にするのをやめて４要素の配列に書き換え。大工事だったけれどコードがすっきり。
- 検索結果上限に達したところで検索を停止するためのモードを新設。
  - （※結果を全保存しておくメリットは検索語の再閲覧や加工。加工や絞り込みについてはその辺りもう少し便利にしていく予定）
- 特に根拠はないけどバージョン0.6

Jan 16, 2012
- .status コマンドを作成。最後に行った検索とその結果について表示
- 特に根拠はないけどバージョン0.5

Jan 15, 2012
- 最後の結果をもう一度見る .last コマンドを実装。
  - .last 3-4 6 9- のように、何番目の結果を表示するか指定することが可能。

Jan 14, 2012
- 検索と表示の所要時間を表示
- 発音記号と用例が（あれば）表示されるようになりました
- 発音記号と用例もインデックス化に対応
- ブログエントリ書いた http://naoyat.hatenablog.jp/entry/2012/01/14/220745
- 英辞郎の辞書をグループ化する際、和英辞郎は抜いてもよさそうだ（全部ロードするとAirには重い）

Jan 13, 2012
- 正規表現検索を実装
  - re2を利用
  - （suffix array検索同様）まだ見出し語一覧しか返していないので、訳語も取り出したい→取り出せるようになった
  - 結果が数千〜数万レコードになる場合、訳語の取り出しは不要な場合もあるので、on/offできるように
- PDIC6の見出し語インデックスの文字変換に対応
  - ABC → abc
  - I'm → i m
  - ジャパン → シヤハン
    - 「パーラー」を検索すると「ハーラー」も出てくるのでイマイチ。元の検索語で結果をチェックする必要がある
- PDIC辞書ファイルもfopen()ではなくmmap()で扱うように変更。
- 結果出力中の検索文字列を強調表示(ANSI color)。re2のGlobalReplaceを使っているので正規表現検索の結果にも対応。
- 見出し語だけでなく訳語についてもインデックス＆suffix arrayを構築し活用。

Jan 12, 2012
- suffix arrayを利用した見出し語の高速全探索を実装
  - シェルから "*word" で呼び出せるようにした
  - .make index でインデックスを作成。あとは辞書ロード時にmmapで読み込まれる
  - 英単語の途中からは検索できない（というか切れ目以外をインデックスに載せてない）けどどうだろう
  - 英語でない場合はUTF-8の１文字単位で（のはずだけどまだノーチェック）

Jan 11, 2012
- suffix arrayを実装
  - 見出し語の途中からマッチするものを高速に取り出したり
  - ディスクやメモリに余裕のある人は訳語全文からの検索にも使いたい
  - まだ使われていない
    - 無ければその場で（といってもちょっと時間かかるよ）作ってファイルに書き出し。mmapして使う感じ
    - 辞書ファイルのloadpathにあれば認識してくれる、で良いかな
- シェル機能を分離
  - lookup("word") の結果を vector<pair<string,string> > で返すように変更。ユニットテストしやすくした
  - 見た目は同一文字列なのに EXPECT_STREQ() が false になるのでおかしいなあと思ってダンプしてみたら、全角チルダが英辞郎の内部で U+FF5E(FULLWIDTH TILDE) になっている。コピペしても U+301C(WAVE DASH) にされてしまうMacでどうしろと

Jan 10, 2012
- clock()とgettimeofday()を併用して、プロセス時間と実時間を得る
- コードちょっと整理

Jan 9, 2012
- サロゲートペア（の変換）に対応
- コードちょっと整理
- コールバックに渡すデータをできるだけ未加工なものにしたら1秒未満で舐められた
  - 各種変換の前準備のためにPASS1で一回舐めておきたいのです

Jan 8, 2012
- 英辞郎131の全単語を舐めるのにうちのAirでプロセス時間3秒ちょい（実時間10秒弱？）かかってる
  - 圧縮見出し語を解凍するのにmallocしたりmemcpyしたり、BOCU-1からutf8に変換したりを抜いたらもっと短いかと
- GNU LGPL ver.2採用

Jan 7, 2012
- ~/.pdicrc ファイルがあれば読む
- 別名(alias)を付与
- 辞書グループ(group)串刺し検索
- PDIC ver 4.00 のデータもSJISなら見られる（けどfr.dicとか無限ループしてる）
- 単語全ダンプとか
- 発音記号や用例をまだ出してない

Jan 6, 2012
- 作り始め
- written in GNU G++
- とりあえず Mac OS X 上でしか見てない
- とりあえず英辞郎131しか見てない
- とりあえず PDIC ver 6.10 しか見てない


bocu1.cc
#include "bocu1.h"

#include <cstdio>
#include <cstdlib>
#include <cstring>
#include "util.h"
#include "dump.h"

int decode_trail[256] = {
   -1,   0,   1,   2,   3,   4,   5,  -1,  -1,  -1,  -1,  -1,  -1,  -1,  -1,  -1,
    6,   7,   8,   9,  10,  11,  12,  13,  14,  15,  -1,  -1,  16,  17,  18,  19,
   -1,  20,  21,  22,  23,  24,  25,  26,  27,  28,  29,  30,  31,  32,  33,  34,
   35,  36,  37,  38,  39,  40,  41,  42,  43,  44,  45,  46,  47,  48,  49,  50,
   51,  52,  53,  54,  55,  56,  57,  58,  59,  60,  61,  62,  63,  64,  65,  66,
   67,  68,  69,  70,  71,  72,  73,  74,  75,  76,  77,  78,  79,  80,  81,  82,
   83,  84,  85,  86,  87,  88,  89,  90,  91,  92,  93,  94,  95,  96,  97,  98,
   99, 100, 101, 102, 103, 104, 105, 106, 107, 108, 109, 110, 111, 112, 113, 114,
  115, 116, 117, 118, 119, 120, 121, 122, 123, 124, 125, 126, 127, 128, 129, 130,
  131, 132, 133, 134, 135, 136, 137, 138, 139, 140, 141, 142, 143, 144, 145, 146,
  147, 148, 149, 150, 151, 152, 153, 154, 155, 156, 157, 158, 159, 160, 161, 162,
  163, 164, 165, 166, 167, 168, 169, 170, 171, 172, 173, 174, 175, 176, 177, 178,
  179, 180, 181, 182, 183, 184, 185, 186, 187, 188, 189, 190, 191, 192, 193, 194,
  195, 196, 197, 198, 199, 200, 201, 202, 203, 204, 205, 206, 207, 208, 209, 210,
  211, 212, 213, 214, 215, 216, 217, 218, 219, 220, 221, 222, 223, 224, 225, 226,
  227, 228, 229, 230, 231, 232, 233, 234, 235, 236, 237, 238, 239, 240, 241, 242
};

int encode_trail[243] = {
    1,   2,   3,   4,   5,   6,  16,  17,  18,  19,  20,  21,  22,  23,  24,  25,
   28,  29,  30,  31,  33,  34,  35,  36,  37,  38,  39,  40,  41,  42,  43,  44,
   45,  46,  47,  48,  49,  50,  51,  52,  53,  54,  55,  56,  57,  58,  59,  60,
   61,  62,  63,  64,  65,  66,  67,  68,  69,  70,  71,  72,  73,  74,  75,  76,
   77,  78,  79,  80,  81,  82,  83,  84,  85,  86,  87,  88,  89,  90,  91,  92,

   93,  94,  95,  96,  97,  98,  99, 100, 101, 102, 103, 104, 105, 106, 107, 108,
  109, 110, 111, 112, 113, 114, 115, 116, 117, 118, 119, 120, 121, 122, 123, 124,
  125, 126, 127, 128, 129, 130, 131, 132, 133, 134, 135, 136, 137, 138, 139, 140,
  141, 142, 143, 144, 145, 146, 147, 148, 149, 150, 151, 152, 153, 154, 155, 156,
  157, 158, 159, 160, 161, 162, 163, 164, 165, 166, 167, 168, 169, 170, 171, 172,

  173, 174, 175, 176, 177, 178, 179, 180, 181, 182, 183, 184, 185, 186, 187, 188,
  189, 190, 191, 192, 193, 194, 195, 196, 197, 198, 199, 200, 201, 202, 203, 204,
  205, 206, 207, 208, 209, 210, 211, 212, 213, 214, 215, 216, 217, 218, 219, 220,
  221, 222, 223, 224, 225, 226, 227, 228, 229, 230, 231, 232, 233, 234, 235, 236,
  237, 238, 239, 240, 241, 242, 243, 244, 245, 246, 247, 248, 249, 250, 251, 252,
  253, 254, 255
};

unsigned char *encode_bocu1(unichar *src_codepoint, int src_len, int& dest_size)
{
  dest_size = 0;
  unsigned char* dest = (unsigned char *)malloc(4*src_len+1);

  for (int i=0,pc=0x40; i<src_len; ) {
    int code = src_codepoint[i++], diff = code - pc;

    if (code <= 0x20) {
      dest[dest_size++] = code;
    } else if (diff < -14536567) {
      // underflow exception
    } else if (diff < -187660) {
      diff -= -14536567;
      int t3 = diff % 243; diff /= 243;
      int t2 = diff % 243; diff /= 243;
      int t1 = diff % 243; diff /= 243;
      // int t0 = diff; // = 0
      dest[dest_size++] = 0x21;
      dest[dest_size++] = encode_trail[t1];
      dest[dest_size++] = encode_trail[t2];
      dest[dest_size++] = encode_trail[t3];
    } else if (diff < -10513) {
      diff -= -187660;
      int t2 = diff % 243; diff /= 243;
      int t1 = diff % 243; diff /= 243;
      int t0 = diff;
      dest[dest_size++] = 0x22 + t0;
      dest[dest_size++] = encode_trail[t1];
      dest[dest_size++] = encode_trail[t2];
    } else if (diff < -64) {
      diff -= -10513;
      int t1 = diff % 243; diff /= 243;
      int t0 = diff;
      dest[dest_size++] = 0x25 + t0;
      dest[dest_size++] = encode_trail[t1];
    } else if (diff < 64) {
      diff -= -64;
      int t0 = diff;
      dest[dest_size++] = 0x50 + t0;
    } else if (diff < 10513) {
      diff -= 64;
      int t1 = diff % 243; diff /= 243;
      int t0 = diff;
      dest[dest_size++] = 0xd0 + t0;
      dest[dest_size++] = encode_trail[t1];
    } else if (diff < 187660) {
      diff -= 10513;
      int t2 = diff % 243; diff /= 243;
      int t1 = diff % 243; diff /= 243;
      int t0 = diff;
      dest[dest_size++] = 0xfb + t0;
      dest[dest_size++] = encode_trail[t1];
      dest[dest_size++] = encode_trail[t2];
    } else if (diff < 14536567) {
      diff -= 187660;
      int t3 = diff % 243; diff /= 243;
      int t2 = diff % 243; diff /= 243;
      int t1 = diff % 243; diff /= 243;
      // int t0 = diff; // = 0
      dest[dest_size++] = 0xfe;
      dest[dest_size++] = encode_trail[t1];
      dest[dest_size++] = encode_trail[t2];
      dest[dest_size++] = encode_trail[t3];
    } else {
      // overflow exception
    }

    // renew pc
    if (code < 0x20) pc = 0x40;
    else if (code == 0x20) ;
    else if (0x3040 <= code && code <= 0x309f) pc = 0x3070;
    else if (0x4e00 <= code && code <= 0x9fa5) pc = 0x7711;
    else if (0xac00 <= code && code <= 0xd7a3) pc = 0xc1d1;
    else pc = (code & 0xffff80) | 0x40;
  }
  dest[dest_size] = 0;

  unsigned char *newp = (unsigned char *)realloc((void *)dest, dest_size+1);
  return newp ? newp : dest;
}

unichar *decode_bocu1(unsigned char *src_bocu1, int src_size, int& dest_size)
{
  dest_size = 0;
  unichar* dest = (unichar *)malloc(sizeof(unichar)*(src_size+1));
  for (int i=0,pc=0x40,lead=0,tr=0,diff=0; i<src_size; ) {
    lead = (unsigned char)src_bocu1[i++];
    if (lead <= 0x20) {
      // code = lead
    }
    else if (lead == 0x21) { // 21 (L T T T)
      diff = -187660 + 243*243*243;
      tr = decode_trail[ src_bocu1[i++] ];
      diff += tr * 243*243;
      tr = decode_trail[ src_bocu1[i++] ];
      diff += tr * 243;
      tr = decode_trail[ src_bocu1[i++] ];
      diff += tr;
    }
    else if (lead < 0x25) { // 22-24 (L T T)
      diff = -10513 + (lead - 0x25)*243*243;
      tr = decode_trail[ src_bocu1[i++] ];
      diff += tr * 243;
      tr = decode_trail[ src_bocu1[i++] ];
      diff += tr;
    }
    else if (lead < 0x50) { // 25-4f (L T)
      diff = -64 + (lead - 0x50)*243;
      tr = decode_trail[ src_bocu1[i++] ];
      diff += tr;
    }
    else if (lead < 0xd0) { // 50-cf (L)
      diff = lead - 0x90;
    }
    else if (lead < 0xfb) { // d0-fa (L T)
      diff = 64 + (lead - 0xd0)*243;
      tr = decode_trail[ src_bocu1[i++] ];
      diff += tr;
    }
    else if (lead < 0xfe) { // fb-fd (L T T)
      diff = 10513 + (lead - 0xfb)*243*243;
      tr = decode_trail[ src_bocu1[i++] ];
      diff += tr * 243;
      tr = decode_trail[ src_bocu1[i++] ];
      diff += tr;
    }
    else if (lead == 0xfe) { // fe (L T T T)
      diff = 187660;
      tr = decode_trail[ src_bocu1[i++] ];
      diff += tr * 243*243;
      tr = decode_trail[ src_bocu1[i++] ];
      diff += tr * 243;
      tr = decode_trail[ src_bocu1[i++] ];
      diff += tr;
    }
    else { // ff : reset
      ;
    }

    if (lead < 0x20) {
      unichar code = (unichar)lead;
      dest[dest_size++] = code;
      pc = 0x40;
    }
    else if (lead == 0x20) {
      unichar code = 0x20;
      dest[dest_size++] = code;
    } else if (lead < 0xff) {
      unichar code = (unichar)(pc + diff);
      //if (code < 0) code = 0; // error recovery
      dest[dest_size++] = code;
      if (code < 0x20) pc = 0x40;
      else if (code == 0x20) ;
      else if (0x3040 <= code && code <= 0x309f) pc = 0x3070;
      else if (0x4e00 <= code && code <= 0x9fa5) pc = 0x7711;
      else if (0xac00 <= code && code <= 0xd7a3) pc = 0xc1d1;
      else pc = (code & ~0x7f) | 0x40;
    } else {
      pc = 0x40; // reset
    }
  }
  dest[dest_size] = 0;

  unichar *newp = (unichar *)realloc((void *)dest, sizeof(unichar)*(dest_size+1));
  return newp ? newp : dest;
}

unsigned char *utf8_to_bocu1(unsigned char *src_utf8, int src_size)
{
  if (!src_size) src_size = strlen((char *)src_utf8);
                     
  int codepoints_len;
  unichar* codepoints = decode_utf8(src_utf8, src_size, codepoints_len);

  int dest_size;
  unsigned char* dest_bocu1 = encode_bocu1(codepoints, codepoints_len, dest_size);

  free((void *)codepoints);
  return dest_bocu1;
}

unsigned char *bocu1_to_utf8(unsigned char *src_bocu1, int src_size)
{
  if (!src_size) src_size = strlen((char *)src_bocu1);

  int codepoints_len;
  unichar* codepoints = decode_bocu1(src_bocu1, src_size, codepoints_len);

  int dest_size;
  unsigned char* dest_utf8 = encode_utf8(codepoints, codepoints_len, dest_size);

  free((void *)codepoints);
  return dest_utf8;
}




void bocu1_check(unsigned char *bocu1_encoded_data, int size)
{
  if (!size) size = strlen((char *)bocu1_encoded_data);

  printf("["); inline_dump(bocu1_encoded_data, size); printf("]");

  int codepoints1_len;
  unichar *codepoints1 = decode_bocu1(bocu1_encoded_data, size, codepoints1_len);
  printf("\n => {"); inline_dump16(codepoints1, codepoints1_len); printf("} => \n");

  int utf8_len;
  unsigned char *utf8str = encode_utf8(codepoints1, codepoints1_len, utf8_len);
  printf(" => \"%s\" [", utf8str); inline_dump(utf8str, utf8_len); printf("]");

  int codepoints2_len;
  unichar *codepoints2 = decode_utf8(utf8str, utf8_len, codepoints2_len);
  printf(" => {"); inline_dump16(codepoints2, codepoints2_len); printf("}");

  int bocu1_len;
  unsigned char *bocu1str = encode_bocu1(codepoints2, codepoints2_len, bocu1_len);
  printf("["); inline_dump(bocu1str, bocu1_len); printf("]");

  if (codepoints2_len == codepoints1_len && memcmp((char *)codepoints1, (char *)codepoints2, sizeof(unichar)*codepoints1_len) == 0)
    printf(" => ok");
  else
    printf(" => NG");

  if (bocu1_len == size && memcmp((char *)bocu1str, (char *)bocu1_encoded_data, size) == 0)
    printf(" => ok.\n");
  else
    printf(" => NG.\n");

  free((void *)utf8str);
  free((void *)codepoints1);
  free((void *)codepoints2);
  free((void *)bocu1str);
}
